spec/index.bs

```diff
@@ -1330,29 +1330,42 @@ enum GPUInputStepMode {
 </script>
 
 <script type=idl>
-dictionary GPUVertexAttributeDescriptor {
-    GPUBufferSize offset = 0;
-    required GPUVertexFormat format;
-    required unsigned long shaderLocation;
+dictionary GPUVertexInputDescriptor {
+    GPUIndexFormat indexFormat = "uint32";
+    sequence<GPUVertexBufferDescriptor?> vertexBuffers = [];
 };
 </script>
 
+A <dfn dfn>vertex buffer</dfn> is a view into buffer memory as an *array of structures*.
+{{GPUVertexBufferDescriptor/elementStride}} is the stride, in bytes, between *elements* of the array.
+Each element of a vertex buffer is a *structure* whose memory layout is defined by
+{{GPUVertexBufferDescriptor/structure}}, which describes the *members* of the structure.
+
+Each {{GPUVertexStructureMemberDescriptor}} describes its
+{{GPUVertexStructureMemberDescriptor/format}} and its
+{{GPUVertexStructureMemberDescriptor/offset}}, in bytes, within the structure.
+
+When used in a vertex shader, members are accessed by their *location*,
+which is specified by {{GPUVertexStructureMemberDescriptor/shaderLocation}}.
+All of the locations must be unique within the {{GPUVertexInputDescriptor}}.
+
 <script type=idl>
 dictionary GPUVertexBufferDescriptor {
-    required GPUBufferSize stride;
+    required GPUBufferSize elementStride;
     GPUInputStepMode stepMode = "vertex";
-    required sequence<GPUVertexAttributeDescriptor> attributeSet;
+    required sequence<GPUVertexStructureMemberDescriptor> structure;
```

Comment:
There might be a bit of a clash between "element" and "structure". Perhaps, we can use one of those more consistently, e.g. call the other thing `structureStride`, or rename this one to `sequence<GPUVertexElementDescriptor> elements`?

Reply:
I was going to comment on this, but it got a bit too close to bikeshedding, but I think this exposes the problem.

Structures aren't the same as elements. The mental model here is: ```c
struct {
    float3 Position;
    float2 TexCoord;
    float4 Color;
} VertexArray[];
```

VertexArray[i] is an "element", VertexArray[i].Position is a "structure member".

Ultimately, I don't like the name "element" because it could also mean "structure element" (as in structure member), or "primitive element" (e.g. one triangle, as used by glDrawElements)

This also ignores the rare but somewhat common case that vertex arrays might have extra data used by the draw, and that's OK. That is:

```
struct {
    float4 Position;
    char pad[29];
} VertexArray[];
```

is perfectly allowed just by setting the buffer stride to a greater value, and using a single element at offset +0x00. Imagine if you could explicitly size a struct with `__attribute__((size=0x3f)) struct Foo` or something in GCC.

Reply:
Could you distill that to concrete wording and names you think we should be using? Or are you arguing that the PR as it stands is clear enough?

Reply:
Quick survey:

## D3D12

* `GPUVertexInputDescriptor` maps to an "input layout" (described by [D3D12_INPUT_LAYOUT_DESC](https://docs.microsoft.com/en-us/windows/win32/api/d3d12/ns-d3d12-d3d12_input_layout_desc)).

* `GPUVertexBufferDescriptor` is the odd one out, since it's not required at input layout creation time. Instead, it's part of [https://docs.microsoft.com/en-us/windows/win32/api/d3d12/ns-d3d12-d3d12_vertex_buffer_view](D3D12_VERTEX_BUFFER_VIEW), which also includes the buffer handle for the data.

* `GPUVertexAttributeDescriptor` maps to [https://docs.microsoft.com/en-us/windows/win32/api/d3d12/ns-d3d12-d3d12_input_element_desc](D3D12_INPUT_ELEMENT_DESC) (another mismatch for "element", being used in the "structure element" case as above). Instead of it being hierarchical, a number-based binding scheme is used.

## Vulkan

* `GPUVertexInputDescriptor` maps to [VkPipelineVertexInputStateCreateInfo](https://www.khronos.org/registry/vulkan/specs/1.1-extensions/man/html/VkPipelineVertexInputStateCreateInfo.html)

* `GPUVertexBufferDescriptor` maps to [VkVertexInputBindingDescription](https://www.khronos.org/registry/vulkan/specs/1.1-extensions/man/html/VkVertexInputBindingDescription.html), again, though, a number-based binding is used like D3D11/D3D12 rather than the hierarchy.

* `GPUVertexAttributeDescriptor` maps to [VkVertexInputAttributeDescription](https://www.khronos.org/registry/vulkan/specs/1.1-extensions/man/html/VkVertexInputAttributeDescription.html)

## Jasper's Personal Taste

I'm not entirely happy with "GPUVertexBufferDescriptor", because it's not *yet* attached to a vertex buffer with data, it's simply a skeleton for what might fit. However, I'm having a hard time coming up with something better, especially if we don't have the number-based binding approach and mandate strides at input layout creation time. "VertexInputBinding" is way worse.

GPUVertexInputDescriptor and GPUVertexAttributeDescriptor are fine.

Reply:
> got a bit too close to bikeshedding

I think this is 100% ok. After all, this PR is pretty much just renaming things.

I will catch up on this thread when I have a chance.

Reply:
@magcius thank you for the mini-investigation!
I think the links you provided support the idea of *not* using the "element" as this PR suggests: D3D12 "element" is a member of the struct (so, our "struct member"). I don't think we should introduce semantics that will confuse D3D12 users, if we can avoid this.

How about "elementStride" -> "structureStride" and "structure" -> "structureMembers"?

Reply:
That sounds pretty good to me. Maybe `arrayStride` and `structureMembers`?

> I'm not entirely happy with "GPUVertexBufferDescriptor"

Perhaps we should steal from Vulkan: GPUVertexBindingDescriptor?

Reply:
Updated with arrayStride and structMembers